### PR TITLE
New optimizer for L1 and L2 penalty

### PR DESCRIPTION
**Pull request by mauriceweiler**
The new optimizer can handle L1 and L2 penalties. AnnealedAdam now inherits from this optimizer.

### CODE REVIEW

**Comment by nasimrahaman** [comment]
Thanks @mauriceweiler! 

Few remarks: 

- I think we should consolidate all `Adam` classes to one. 
- there's a [learning rate scheduling API](https://github.com/pytorch/pytorch/pull/1370) merged in Pytorch master. Maybe we should integrate that somehow?

**Comment by mauriceweiler** [comment]
Definitively a good idea to unify everything. The additional L1 penalty could be easily included in all optimizers in the pytorch master.
Given that the learning rate scheduling API supports multiple means of decaying we should adopt it.

**Comment by nasimrahaman** [comment]
Regarding L1 penality - agreed. Do you want to open an issue/PR at Pytorch?

**Comment by nasimrahaman** [comment]
Had another good look at the LR-Scheduler API. It appears that we have two options: 

- integrate it directly in `inferno.trainers.basic.Trainer`, i.e. we would need a `Trainer.build_learning_rate_scheduler(...)` method, analogous to [`Trainer.build_optimizer(...)`](https://github.com/nasimrahaman/inferno/blob/master/inferno/trainers/basic.py#L174) and [`Trainer.build_criterion(...)`](https://github.com/nasimrahaman/inferno/blob/master/inferno/trainers/basic.py#L227). But once that's out of the way, it should be easy to integrate in `Trainer`'s [`Trainer.next_epoch()`](https://github.com/nasimrahaman/inferno/blob/master/inferno/trainers/basic.py#L662) method. 
__Pros__: we might be able to pickle some schedulers - atleast the ones that don't contain function handles. 
__Cons__: adds complexity to the already complex `Trainer` class. 

- have the scheduler as a callback. The scheduler's step method is [called after every epoch](https://github.com/Jiaming-Liu/pytorch/blob/db59187201d18c203f084c8b288e07f7bd2a43aa/torch/optim/lr_scheduler.py#L26), so all we need to do is to write a `end_of_epoch` callback ([see this](https://github.com/nasimrahaman/inferno/blob/master/inferno/trainers/callbacks/essentials.py#L7) for an example of a `end_of_training_iteration` callback). 
__Pros__: Simple, easily integrates with the native pytorch API for LR-Scheduling, and IMHO elegant. 
__Cons__: Can't be pickled (yet).

**Comment by mauriceweiler** [comment]
Yeah, we should got with p.data.sign()

Regarding the lr scheduler I can't say how important it is to be able to pickle them. Sticking with the native pytorch API is a good idea.

I added a data_slice option to the TIFVolumeLoader which induced a redundancy in the parse_data_slice method - maybe we should define it somewhere externally, e.g. in the utils?

**Comment by nasimrahaman** [comment]
Regarding `parse_data_slice`: good idea. Let's discuss this tomorrow before merging.
